### PR TITLE
Implement todo app

### PR DESCRIPTION
**Pull request by veronicazh**

### CODE REVIEW

File: TodoMain/StatusBar/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+export default observer(function StatusBar ({ todos, $todos }) {
+  // let [isDoneArray, setIsDoneArray] = useState([])
+  let [total, $total] = useQuery('todos', { $count: true })
+  let [isDone, $isDone] = useQuery('todos', { isDone: true })
+  return pug`
+    View.container
+      View.counterContainer
+        Text.text Total: #{total}
+      View.counterContainer
+        Text.text Done: #{isDone.length}
```
Comment (oesnet) [inline TodoMain/StatusBar/index.js]: @veronicazh $count

File: TodoMain/TodoInput/index.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import './index.styl'
+import {
+  View,
+  Text,
+  TextInput,
+  TouchableOpacity
+} from 'react-native'
+import { observer } from 'startupjs'
+
+export default observer(function TodoInput ({ inputValue, setInputValue, addTodo }) {
```
Comment (oesnet) [inline TodoMain/TodoInput/index.js]: @veronicazh AddTodoForm

Comment (oesnet) [reply TodoMain/TodoInput/index.js]: @veronicazh please use local state to handle a value from the input.

Comment (oesnet) [reply TodoMain/TodoInput/index.js]: @veronicazh addTodo -> onSave (onAdd)

File: TodoMain/TodoInput/index.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  function onInputChange (text) {
+    setInputValue(text)
+  }
+
+  return pug`
+    View.root
+      Text
```
Comment (oesnet) [inline TodoMain/TodoInput/index.js]: @veronicazh ?

File: TodoMain/TodoInput/index.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      View.inputContainer
+        TextInput.textInput(
+          value=inputValue
+          onChangeText=(text) => onInputChange(text)
+        )
+        TouchableOpacity.add(
+          onPress=() => addTodo()
```
Comment (oesnet) [inline TodoMain/TodoInput/index.js]: @veronicazh  => onSave(inputValue)

File: TodoMain/TodoList/index.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import React, { useState } from 'react'
+import './index.styl'
+import { observer } from 'startupjs'
+import {
+  View,
+  Text,
+  TouchableOpacity,
+  TextInput
+} from 'react-native'
+import Modal from '../Modal'
```
Comment (oesnet) [inline TodoMain/TodoList/index.js]: @veronicazh please try 'components/Modal' (Don't forget to move Modal to components folder)

File: TodoMain/TodoList/index.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+export default observer(function TodoList ({ todos, toggleImportant, toggleDone, deleteItem, newInputValue, setNewInputValue, editItem }) {
```
Comment (oesnet) [inline TodoMain/TodoList/index.js]: @veronicazh please move whole business logic from the parent here.

File: TodoMain/index.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import StatusBar from './StatusBar'
+
+export default observer(function TodoMain () {
+  let [inputValue, setInputValue] = useState('')
+  let [newInputValue, setNewInputValue] = useState('')
+  let [todos, $todos] = useQuery('todos', {})
+
+  function addTodo () {
+    let itemObject = {
+      title: inputValue,
```
Comment (oesnet) [inline TodoMain/index.js]: @veronicazh please trim incoming text

File: Root.js [hunk 8]
```diff
@@ -1,87 +1,44 @@
...
-    setStateCounter(stateCounter - 1)
-  }
-
-  async function reset () {
-    $counter.reset() // custom ORM method (see /model/)
-    setStateCounter(0)
-  }
-
+  const base = BASE_URL
+  console.log(BASE_URL, 'BASE URL')
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply Root.js]: done

File: Root.js [hunk 9]
```diff
@@ -1,87 +1,44 @@
...
-      TouchableOpacity.button.clear(onPress=reset)
-        Text.label RESET
-      Text.api /api (updated each 3 sec): #{JSON.stringify(api)}
+    ScrollView
+      View.wrapper
+        View.root
+          View.logosContainer
+            Image.logo(
+              source={uri: base + '/img/logo.png'}
+              style={width: 230, height: 53}
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply Root.js]: done

File: TodoMain/TodoInput/index.js (hunk 3 above)
Comment (veronicazh) [reply TodoMain/TodoInput/index.js]: done

File: TodoMain/index.js (hunk 7 above)
Comment (veronicazh) [reply TodoMain/index.js]: done

File: TodoMain/TodoList/index.js (hunk 5 above)
Comment (veronicazh) [reply TodoMain/TodoList/index.js]: done

File: Root.js [hunk 10]
```diff
@@ -1,87 +1,44 @@
...
-        Text.label +
-      TouchableOpacity.button.decrement(onPress=decrement)
-        Text.label -
-      TouchableOpacity.button.clear(onPress=reset)
-        Text.label RESET
-      Text.api /api (updated each 3 sec): #{JSON.stringify(api)}
+    ScrollView
+      View.wrapper
+        View.root
+          View.logosContainer
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply Root.js]: done

[the remark this answers is not in the document]
Comment (veronicazh) [reply Root.js]: done

File: Root.js [hunk 11]
```diff
@@ -1,87 +1,44 @@
...
-      Text DB Counter: #{counter && counter.value}
-      Text State Counter: #{stateCounter}
-      TouchableOpacity.button.increment(onPress=increment)
-        Text.label +
-      TouchableOpacity.button.decrement(onPress=decrement)
-        Text.label -
-      TouchableOpacity.button.clear(onPress=reset)
-        Text.label RESET
-      Text.api /api (updated each 3 sec): #{JSON.stringify(api)}
+    ScrollView
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply Root.js]: done

File: Root.styl [hunk 12]
```diff
@@ -1,26 +1,18 @@
-.body
-  padding gutter(2)
+.wrapper
+  background-color #ffef3f
+  padding 5px
 
-.greeting
-  color $PRIMARY // use a global variable from /styles/index.styl
+.root
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply Root.styl]: done

File: TodoMain/TodoInput/index.js (hunk 4 above)
Comment (veronicazh) [reply TodoMain/TodoInput/index.js]: done

File: TodoMain/TodoList/index.js (hunk 6 above)
Comment (veronicazh) [reply TodoMain/TodoList/index.js]: done

File: TodoMain/StatusBar/index.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import './index.styl'
+import {
+  View,
+  Text
+} from 'react-native'
+import { observer, useQuery } from 'startupjs'
+
+export default observer(function StatusBar ({ todos, $todos }) {
+  // let [isDoneArray, setIsDoneArray] = useState([])
+  let [total, $total] = useQuery('todos', { $count: true })
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply TodoMain/StatusBar/index.js]: done

File: TodoMain/StatusBar/index.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  let [isDone, $isDone] = useQuery('todos', { isDone: true })
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply TodoMain/StatusBar/index.js]: @oesnet done

File: TodoMain/StatusBar/index.js (hunk 1 above)
Comment (veronicazh) [reply TodoMain/StatusBar/index.js]: @oesnet done

File: TodoMain/TodoInput/index.js (hunk 2 above)
Comment (veronicazh) [reply TodoMain/TodoInput/index.js]: @oesnet done

File: Root.js [hunk 15]
```diff
@@ -1,87 +1,28 @@
...
 import {
-  observer,
-  useDoc,
-  useApi
+  observer
 } from 'startupjs'
-import axios from 'axios'
 import './Root.styl'
 
+import Todos from './screens/Todos'
```
Comment (oesnet) [inline Root.js]: @veronicazh please try `'screens/Todos'`

File: Root.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (oesnet) [reply Root.js]: @veronicazh please use ScrollView in the component.

[the remark this answers is not in the document]
Comment (oesnet) [reply Root.js]: @veronicazh please read https://medium.com/@peterpme/taming-react-natives-scrollview-with-flex-144e6ff76c08

File: components/AddTodoForm/index.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  View,
+  Text,
+  TextInput,
+  TouchableOpacity
+} from 'react-native'
+import { observer, useQuery } from 'startupjs'
+
+export default observer(function AddTodoForm () {
+  let [inputValue, setInputValue] = useState('')
+  let [, $todos] = useQuery('todos', {})
```
Comment (oesnet) [inline components/AddTodoForm/index.js]: @veronicazh please try to useModel('todos') to get just an appropriate collection which hasn't been initialised with data.

File: Root.js (hunk 15 above)
Comment (veronicazh) [reply Root.js]: doesn't work

File: components/List/index.js [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+  function editItem (todoId) {
+    $todos.set(todoId + '.title', newInputValue)
+  }
+
+  return pug`
+    View.root
+      Text.listTitle My tasks:
+      View.list
+      Modal.modal(
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh let's move modal and its content to a separate component `EditTodoModal`

File: components/List/index.js [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        visible=modalVisible
+        transparent=true
+        animationType='fade'
+        ariaHideApp=false
+      )
+        View.modalInner
+          TextInput.inputModal(
+            value=newInputValue
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh it will be a local state of `EditTodoModal `

File: components/List/index.js [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+            onChangeText=(text) => setNewInputValue(text)
+            autoFocus=true
+          )
+          TouchableOpacity.okButton(
+            onPress=() => { setModalVisible(!modalVisible); editItem(activeItemId) }
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh it will be onSave callback

File: components/List/index.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+} from 'react-native'
+import Modal from '../Modal'
+
+export default observer(function List () {
+  let [todos, $todos] = useQuery('todos', {})
+  let [modalVisible, setModalVisible] = useState(false)
+  let [activeItemId, setActiveItemId] = useState('')
+  let [newInputValue, setNewInputValue] = useState('')
+
+  function toggleImportant (todo) {
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh let's receive in argument just id. Then we get current value within the function and invert its value.
const `$todo = $todos.at(todo.id)` - $todo.set([field], [value])

File: components/List/index.js [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    $todos.set(todo.id + '.isImportant', !todo.isImportant)
+  }
+
+  function toggleDone (todo) {
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh same

File: components/List/index.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+    $todos.set(todo.id + '.isDone', !todo.isDone)
+  }
+
+  function deleteItem (todo) {
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh same

File: components/List/index.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+    $todos.del(todo.id)
+  }
+
+  function editItem (todoId) {
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh `id, title`

File: components/List/index.js [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+          )
+          TouchableOpacity.okButton(
+            onPress=() => { setModalVisible(!modalVisible); editItem(activeItemId) }
+          )
+            Text.textModal() OK
+          TouchableOpacity.cancelButton(
+              onPress=() => setModalVisible(false)
+            )
+            Text.textModal() Cancel
+      each item, index in todos
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh please rename `item` to `todo` to let other devs to understand below what it is.

Comment (oesnet) [inline components/List/index.js]: @veronicazh let's remove `index`. we have `item.id`

File: components/List/index.js [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+        React.Fragment(key=item.id)
```
Comment (oesnet) [inline components/List/index.js]: @veronicazh this line is unnecessary except `key` prop (let move it to the element below)

File: components/StatusBar/index.js [hunk 26]
```diff
@@ -0,0 +1,19 @@
+import React from 'react'
+import './index.styl'
+import {
+  View,
+  Text
+} from 'react-native'
+import { observer, useQuery } from 'startupjs'
+
+export default observer(function StatusBar () {
```
Comment (oesnet) [inline components/StatusBar/index.js]: @veronicazh let's rename to Counters

File: components/StatusBar/index.js [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+  let [totalCountTodos] = useQuery('todos', { $count: true })
```
Comment (oesnet) [inline components/StatusBar/index.js]: @veronicazh totalTodos

File: screens/Todos/index.js [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+import './index.styl'
+import { observer } from 'startupjs'
+import {
+  View
+} from 'react-native'
+
+import AddTodoForm from '../../components/AddTodoForm'
+import List from '../../components/List'
+import StatusBar from '../../components/StatusBar'
+import Header from '../../components/Header'
```
Comment (oesnet) [inline screens/Todos/index.js]: @veronicazh please implement index.js in components folder to import necessary components at once.

File: components/AddTodoForm/index.js (hunk 16 above)
Comment (veronicazh) [reply components/AddTodoForm/index.js]: done

File: components/AddTodoForm/index.js [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+
+  function onSave () {
+    let itemObject = {
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/AddTodoForm/index.js]: done

File: components/AddTodoForm/index.js [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+
+  return pug`
+    View.root
+      View.inputContainer
+        TextInput.textInput(
+          value=inputValue
+          onChangeText=(text) => setInputValue(text)
+        )
+        TouchableOpacity.add(
+          onPress=() => onSave()
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/AddTodoForm/index.js]: done

File: components/AddTodoForm/index.js [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+    $todos.add(itemObject)
+    setInputValue('')
+  }
+
+  return pug`
+    View.root
+      View.inputContainer
+        TextInput.textInput(
+          value=inputValue
+          onChangeText=(text) => setInputValue(text)
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/AddTodoForm/index.js]: done

File: components/Header/index.js [hunk 32]
```diff
@@ -0,0 +1,21 @@
+import { BASE_URL } from '@env'
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/Header/index.js]: done

File: components/Header/index.js [hunk 33]
```diff
@@ -0,0 +1,21 @@
...
+import React from 'react'
+import './index.styl'
+import {
+  View,
+  Image
+} from 'react-native'
+import { observer } from 'startupjs'
+
+export default observer(function Header () {
+  const base = BASE_URL
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/Header/index.js]: done

File: components/List/index.js [hunk 34]
```diff
@@ -0,0 +1,80 @@
+import React, { useState } from 'react'
+import './index.styl'
+import { observer, useQuery } from 'startupjs'
+import {
+  View,
+  Text,
+  TouchableOpacity,
+  TextInput
+} from 'react-native'
+import Modal from '../Modal'
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js [hunk 35]
```diff
@@ -0,0 +1,80 @@
...
+
+export default observer(function List () {
+  let [todos, $todos] = useQuery('todos', {})
+  let [modalVisible, setModalVisible] = useState(false)
+  let [activeItemId, setActiveItemId] = useState('')
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+  View,
+  Text,
+  TouchableOpacity,
+  TextInput
+} from 'react-native'
+import Modal from '../Modal'
+
+export default observer(function List () {
+  let [todos, $todos] = useQuery('todos', {})
+  let [modalVisible, setModalVisible] = useState(false)
```
[the remark this answers is not in the document]
Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js (hunk 24 above)
Comment (veronicazh) [reply components/List/index.js]: done

Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js (hunk 25 above)
Comment (veronicazh) [reply components/List/index.js]: done

File: components/StatusBar/index.js (hunk 26 above)
Comment (veronicazh) [reply components/StatusBar/index.js]: done

File: components/StatusBar/index.js (hunk 27 above)
Comment (veronicazh) [reply components/StatusBar/index.js]: done

File: screens/Todos/index.js (hunk 28 above)
Comment (veronicazh) [reply screens/Todos/index.js]: done

File: components/List/index.js (hunk 17 above)
Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js (hunk 18 above)
Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js (hunk 20 above)
Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js (hunk 21 above)
Comment (veronicazh) [reply components/List/index.js]: done

File: components/List/index.js (hunk 22 above)
Comment (veronicazh) [reply components/List/index.js]: done

File: components/EditTodoModal/index.js [hunk 37]
```diff
@@ -0,0 +1,41 @@
...
+import {
+  View,
+  Text,
+  TextInput,
+  TouchableOpacity
+} from 'react-native'
+import { observer, useDoc } from 'startupjs'
+import Modal from 'components/Modal'
+
+export default observer(function EditTodoModal ({ editItemId, setEditItemId }) {
```
Comment (oesnet) [inline components/EditTodoModal/index.js]: @veronicazh id, onClose

File: components/EditTodoModal/index.js [hunk 38]
```diff
@@ -0,0 +1,41 @@
...
+  let [currentTodo, $currentTodo] = useDoc('todos', editItemId)
```
Comment (oesnet) [inline components/EditTodoModal/index.js]: @veronicazh just `todo`

File: components/EditTodoModal/index.js [hunk 39]
```diff
@@ -0,0 +1,41 @@
...
+  let [inputValue, setInputValue] = useState(currentTodo.title)
+  function editItem () {
```
Comment (oesnet) [inline components/EditTodoModal/index.js]: @veronicazh save

File: components/EditTodoModal/index.js [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+    $currentTodo.set('title', inputValue)
```
Comment (oesnet) [inline components/EditTodoModal/index.js]: @veronicazh we call the callback after changing a title

File: components/EditTodoModal/index.js [hunk 41]
```diff
@@ -0,0 +1,41 @@
...
+      ariaHideApp=false
+    )
+      View.modalInner
+        TextInput.inputModal(
+          value=inputValue
+          onChangeText=(text) => setInputValue(text)
+          autoFocus=true
+        )
+        TouchableOpacity.okButton(
+          onPress=() => { setEditItemId(''); editItem() }
```
Comment (oesnet) [inline components/EditTodoModal/index.js]: @veronicazh onPress=save

File: components/EditTodoModal/index.js [hunk 42]
```diff
@@ -0,0 +1,41 @@
...
+        )
+          Text.textModal OK
+        TouchableOpacity.cancelButton(
+            onPress=() => setEditItemId('')
```
Comment (oesnet) [inline components/EditTodoModal/index.js]: @veronicazh onPress=onClose

File: components/EditTodoModal/index.js (hunk 38 above)
Comment (veronicazh) [reply components/EditTodoModal/index.js]: done

File: components/EditTodoModal/index.js (hunk 39 above)
Comment (veronicazh) [reply components/EditTodoModal/index.js]: done

File: components/EditTodoModal/index.js (hunk 40 above)
Comment (veronicazh) [reply components/EditTodoModal/index.js]: done

File: components/EditTodoModal/index.js (hunk 41 above)
Comment (veronicazh) [reply components/EditTodoModal/index.js]: done

File: components/EditTodoModal/index.js (hunk 42 above)
Comment (veronicazh) [reply components/EditTodoModal/index.js]: done

File: components/EditTodoModal/index.js (hunk 37 above)
Comment (veronicazh) [reply components/EditTodoModal/index.js]: done

**Pull request merged**